samples/guide/src/main/java/okhttp3/recipes/kt/WiresharkExample.kt

```diff
@@ -0,0 +1,287 @@
+        val parameter = parameters[0] as String
+
+        if (message == "Produced ClientHello handshake message") {
+          random = readClientRandom(parameter)
+        }
+      }
+    }
+
+    override fun flush() {}
+
+    override fun close() {}
+  }
+
+  private fun readClientRandom(param: String): String? {
+    val matchResult = randomRegex.find(param)
+
+    return if (matchResult != null) {
+      matchResult.groupValues[1].replace(" ", "")
+    } else {
+      null
+    }
+  }
+
+  override fun secureConnectStart(call: Call) {
+    // Register to capture "Produced ClientHello handshake message".
+    currentThread = Thread.currentThread()
+    logger.addHandler(loggerHandler)
+  }
+
+  override fun secureConnectEnd(
+    call: Call,
+    handshake: Handshake?
+  ) {
+    logger.removeHandler(loggerHandler)
+  }
+
+  override fun callEnd(call: Call) {
+    // Cleanup log handler if failed.
+    logger.removeHandler(loggerHandler)
+  }
+
+  override fun connectionAcquired(
+    call: Call,
+    connection: Connection
+  ) {
+    if (random != null) {
+      val sslSocket = connection.socket() as SSLSocket
+      val session = sslSocket.session
+
+      val masterSecretHex = session.masterSecret?.encoded?.toByteString()
+          ?.hex()
+
+      if (masterSecretHex != null) {
+        val keyLog = "CLIENT_RANDOM $random $masterSecretHex"
+
+        if (verbose) {
+          println(keyLog)
+        }
+        logFile.appendText("$keyLog\n")
+      }
+    }
+
+    random = null
+  }
+
+  class Factory(
+    private val logFile: File,
+    private val verbose: Boolean = false
+  ) : EventListener.Factory {
+    override fun create(call: Call): EventListener {
+      return WireSharkKeyLoggerListener(logFile, verbose)
+    }
+  }
+
+  companion object {
+    private lateinit var logger: Logger
+
+    private val SSLSession.masterSecret: SecretKey?
+      get() = javaClass.getDeclaredField("masterSecret")
+          .apply {
+            isAccessible = true
+          }
+          .get(this) as? SecretKey
+
+    val randomRegex = "\"random\"\\s+:\\s+\"([^\"]+)\"".toRegex()
+
+    fun register() {
+      // Enable JUL logging for SSL events, must be activated early or via -D option.
+      System.setProperty("javax.net.debug", "")
+      logger = Logger.getLogger("javax.net.ssl")
+          .apply {
+            level = Level.FINEST
+          }
+    }
+  }
+}
+
+val eventListenerFactory = WireSharkKeyLoggerListener.Factory(File("/tmp/key.log"), verbose = true)
+
+@SuppressSignatureCheck
+class WiresharkExample(private val tlsVersions: List<TlsVersion>) {
+  val connectionSpec =
+    Builder(ConnectionSpec.RESTRICTED_TLS)
+        .tlsVersions(*tlsVersions.toTypedArray())
+        .build()
+
+  var bootstrapClient = OkHttpClient.Builder()
+      .connectionSpecs(listOf(connectionSpec))
+      .addInterceptor(BrotliInterceptor)
+      .eventListenerFactory(eventListenerFactory)
+      .build()
+
+  val dns = DnsOverHttps.Builder()
```

Comment:
strictly necessary? I think this is also a distraction. Also discourages use for private networks.

Reply:
removed.